File: finalyearproject/repos/dkd-platform-libs/sdk/python/src/dkd_platform/idempotency.py
# HAND-AUTHORED platform primitive (NOT dkdgen-generated).
# PL-03 — Idempotency-Key enforcement helper (EF-API-6).
"""Idempotency-Key enforcement for unsafe / money / custody writes.

Canon (EF-API-6): every unsafe/money/custody write carries an `Idempotency-Key` header. This helper
enforces the three branches so a service never re-applies a money/custody effect:

  1. MISSING key on a guarded write         -> reject with HTTP 400.
  2. SAME key + SAME payload (a retry)       -> REPLAY the original stored response verbatim.
  3. SAME key + DIFFERENT payload (misuse)   -> reject with HTTP 409.
  otherwise (first time this key is seen)    -> PROCEED; the caller runs the handler, then records
                                                the response via `commit` so a later retry replays it.

The store is a pluggable `IdempotencyStore` Protocol — back it with the PL-02 inbox table or a
dedicated idempotency table; the helper NEVER hard-wires a database. `payload` is fingerprinted with
SHA-256, so a mismatch is detected without persisting the raw request body.
"""
from __future__ import annotations

import hashlib
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional, Protocol

HEADER = "Idempotency-Key"

# HTTP statuses this helper emits (see errors.py / http_status.py for the full vocabulary).
_MISSING_KEY = 400
_MISMATCH = 409


def fingerprint(payload: bytes) -> str:
    """SHA-256 hex of the raw request body — the stored request identity for replay/mismatch."""
    if isinstance(payload, str):
        payload = payload.encode("utf-8")
    return hashlib.sha256(payload).hexdigest()


@dataclass(frozen=True)
class StoredResponse:
    """The original response captured for a completed idempotent request."""

    request_hash: str
    status: int
    body: bytes = b""
    headers: tuple[tuple[str, str], ...] = ()


class IdempotencyStore(Protocol):
    """Persistence surface for the guard. A PL-02 inbox row or an idempotency table satisfies it.

    `save` MUST be insert-if-absent (e.g. `ON CONFLICT DO NOTHING`) and return False when a
    concurrent request already committed the key — the guard then replays the winner.
    """

    def load(self, key: str) -> Optional[StoredResponse]: ...

    def save(self, key: str, response: StoredResponse) -> bool: ...


class Action(str, Enum):
    PROCEED = "proceed"
    REPLAY = "replay"
    REJECT = "reject"


@dataclass(frozen=True)
class Decision:
    """The guard's verdict. On REPLAY/REJECT the caller returns `status` immediately (with
    `response` on REPLAY); on PROCEED the caller runs the handler and then calls `commit`."""

    action: Action
    status: int = 0
    response: Optional[StoredResponse] = None
    reason: Optional[str] = None


class IdempotencyGuard:
    """Framework-agnostic enforcer. Wire it into a middleware / dependency for guarded routes."""

    def __init__(self, store: IdempotencyStore):
        self._store = store

    def begin(self, key: Optional[str], payload: bytes) -> Decision:
        """Classify an incoming guarded write BEFORE the handler runs."""
        if not key:
            return Decision(Action.REJECT, _MISSING_KEY, reason="missing Idempotency-Key")
        prior = self._store.load(key)
        if prior is None:
            return Decision(Action.PROCEED)
        if prior.request_hash == fingerprint(payload):
            return Decision(Action.REPLAY, prior.status, response=prior)
        return Decision(Action.REJECT, _MISMATCH, reason="Idempotency-Key reused with a different payload")

    def commit(self, key: str, payload: bytes, status: int, body: bytes = b"",
               headers: tuple[tuple[str, str], ...] = ()) -> StoredResponse:
        """Persist the handler's response so a later retry with the same key replays it."""
        stored = StoredResponse(fingerprint(payload), status, body, headers)
        self._store.save(key, stored)
        return stored

File: finalyearproject/repos/dkd-platform-libs/sdk/csharp/Dkd.Platform/Idempotency.cs
// HAND-AUTHORED platform primitive (NOT dkdgen-generated).
// PL-03: Idempotency-Key enforcement for unsafe / money / custody writes (EF-API-6). The helper is
// driver-agnostic: it reaches persistence through IIdempotencyStore, which a service backs with the
// PL-02 inbox table or a dedicated idem table — the SDK never hard-wires a DB. Three canon branches:
//   * MISSING key on an unsafe write            -> 400 (reject)
//   * SAME key + SAME request payload (replay)  -> return the ORIGINAL stored response
//   * SAME key + DIFFERENT request payload       -> 409 (key reuse mismatch)

namespace Dkd.Platform;

using System.Security.Cryptography;
using System.Text;

/// <summary>The captured HTTP response the platform replays for a repeated idempotency key.</summary>
public sealed record IdempotentResponse(int Status, string Body, string ContentType = "application/json");

/// <summary>One persisted idempotency row: the key, a fingerprint of the original request, its response.</summary>
public sealed record IdempotencyRecord(string Key, string RequestFingerprint, IdempotentResponse Response);

/// <summary>
/// Pluggable idempotency store. Back it with the PL-02 inbox (<see cref="Inbox"/>) or a dedicated
/// <c>idempotency</c> table — the SDK stays DB-agnostic. Implementations MUST persist and look up by
/// the exact <c>Idempotency-Key</c> string.
/// </summary>
public interface IIdempotencyStore
{
    /// <summary>Return the stored record for <paramref name="key"/>, or null when the key is unseen.</summary>
    IdempotencyRecord? Find(string key);

    /// <summary>Persist the first-seen record (key + request fingerprint + captured response).</summary>
    void Save(IdempotencyRecord record);
}

/// <summary>What the caller must do next after evaluating an idempotency key.</summary>
public enum IdempotencyOutcome
{
    /// <summary>No key supplied on an unsafe write — reject with 400.</summary>
    MissingKey,

    /// <summary>Key + payload match a prior request — return <see cref="IdempotencyDecision.Response"/> verbatim.</summary>
    Replay,

    /// <summary>Key reused with a different payload — reject with 409.</summary>
    Conflict,

    /// <summary>First use of this key — run the handler, then call <see cref="Idempotency.Remember"/>.</summary>
    Proceed,
}

/// <summary>The evaluation result. <see cref="HttpStatus"/> is 0 for <see cref="IdempotencyOutcome.Proceed"/>.</summary>
public sealed record IdempotencyDecision(
    IdempotencyOutcome Outcome, int HttpStatus, string? Code, IdempotentResponse? Response);

/// <summary>
/// Stateless idempotency policy. Wrap the three unsafe-write branches of EF-API-6 around a
/// pluggable <see cref="IIdempotencyStore"/>. Typical use:
/// <code>
/// var d = Idempotency.Evaluate(store, req.Header("Idempotency-Key"), rawBody);
/// switch (d.Outcome) {
///   case IdempotencyOutcome.MissingKey: return Problem(400, d.Code);
///   case IdempotencyOutcome.Conflict:   return Problem(409, d.Code);
///   case IdempotencyOutcome.Replay:     return Replay(d.Response!);
///   case IdempotencyOutcome.Proceed:
///     var resp = Handle();
///     Idempotency.Remember(store, key, rawBody, resp);
///     return resp;
/// }
/// </code>
/// </summary>
public static class Idempotency
{
    public const string HeaderName = "Idempotency-Key";
    public const string MissingKeyCode = "dokandar.platform.idempotency.missing_key";
    public const string MismatchCode = "dokandar.platform.idempotency.key_reuse_mismatch";

    /// <summary>SHA-256 (lowercase hex) fingerprint of the raw request payload — the "same payload" key.</summary>
    public static string Fingerprint(string? requestPayload) =>
        Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(requestPayload ?? string.Empty)))
               .ToLowerInvariant();

    /// <summary>Classify an unsafe write against the store per EF-API-6 (does not mutate the store).</summary>
    public static IdempotencyDecision Evaluate(IIdempotencyStore store, string? idempotencyKey, string? requestPayload)
    {
        ArgumentNullException.ThrowIfNull(store);

        if (string.IsNullOrWhiteSpace(idempotencyKey))
            return new IdempotencyDecision(IdempotencyOutcome.MissingKey, 400, MissingKeyCode, null);

        var existing = store.Find(idempotencyKey);
        if (existing is null)
            return new IdempotencyDecision(IdempotencyOutcome.Proceed, 0, null, null);

        if (existing.RequestFingerprint == Fingerprint(requestPayload))
            return new IdempotencyDecision(IdempotencyOutcome.Replay, existing.Response.Status, null, existing.Response);

        return new IdempotencyDecision(IdempotencyOutcome.Conflict, 409, MismatchCode, null);
    }

    /// <summary>Persist the first-seen key + payload fingerprint + response, so later replays return it.</summary>
    public static void Remember(IIdempotencyStore store, string idempotencyKey, string? requestPayload,
        IdempotentResponse response)
    {
        ArgumentNullException.ThrowIfNull(store);
        ArgumentNullException.ThrowIfNull(response);
        Outbox.Require(idempotencyKey, nameof(idempotencyKey));
        store.Save(new IdempotencyRecord(idempotencyKey, Fingerprint(requestPayload), response));
    }
}
